onnxruntime/test/providers/cpu/tensor/cast_op_test.cc

```diff
@@ -207,6 +207,1036 @@ TEST(CastOpTest, ToString) {
   TestCastOp(gsl::make_span(int_16_input), gsl::make_span(int_string_data), shape);
 }
 
+TEST(CastOpTest, Int4x2ToInt8) {
+  // GIVEN
+  const std::vector<int64_t> shape{2, 2, 2};
+  const std::vector<Int4x2> int4x2_input = {
+      Int4x2(-8, 7),  // boundary values
+      Int4x2(0, -1),  // zero and negative
+      Int4x2(3, -5),  // positive and negative
+      Int4x2(6, 2)    // both positive
+  };
+
+  const std::vector<int8_t> expected_int8_output = {-8, 7, 0, -1, 3, -5, 6, 2};
+
+  // WHEN, THEN
+  TestCastOp(gsl::make_span(int4x2_input), gsl::make_span(expected_int8_output), shape);
+}
+
+TEST(CastOpTest, Int4x2ToUInt8) {
+  // GIVEN
+  const std::vector<int64_t> shape{2, 2, 2};
+  const std::vector<Int4x2> int4x2_input = {
+      Int4x2(-8, 7),  // boundary values
+      Int4x2(0, -1),  // zero and negative
+      Int4x2(3, -5),  // positive and negative
+      Int4x2(6, 2)    // both positive
+  };
+
+  // Negative values will be cast to their unsigned representation
+  const std::vector<uint8_t> expected_uint8_output = {248, 7, 0, UINT8_MAX, 3, 251, 6, 2};
+
+  // WHEN, THEN
+  TestCastOp(gsl::make_span(int4x2_input), gsl::make_span(expected_uint8_output), shape);
+}
+
+TEST(CastOpTest, Int4x2ToInt16) {
+  // GIVEN
+  const std::vector<int64_t> shape{2, 2, 2};
```

Comment:
Could you add a test of shape like {3, 1, 2} and {1, 2, 3}? I think uint4x2 and int4x2 has some odd implementation for packing.